### PR TITLE
Avr

### PR DESCRIPTION
**Pull request by tballmsft**
- refactoring to support AVR assembler and Thumb Assembler
- refactoring of backthumb.ts to backbase.ts, hexfile.ts, backthumb.ts

### CODE REVIEW

File: cli/cli.ts [hunk 1]
```diff
@@ -3240,6 +3251,7 @@ cmd("extract  [FILENAME]          - extract sources from .hex/.jsz file, stdin (
 cmd("test                         - run tests on current package", testAsync, 1)
 cmd("gendocs                      - build current package and its docs", gendocsAsync, 1)
 cmd("format   [-i] file.ts...     - pretty-print TS files; -i = in-place", formatAsync, 1)
+cmd("testassembler                - test the assemblers", testAssemblers)
```
Comment (mmoskal) [inline cli/cli.ts]: Needs to have `,1)` or `,2)` to hide from simple help

File: pxtlib/emitter/assembler.ts [hunk 2]
```diff
@@ -323,18 +316,22 @@ namespace ts.pxtc.assembler {
...
+            // stack-specific processing
+
             // more special characters to handle
             if (s.indexOf("@") >= 0) {
                 m = /^(\w+)@(-?\d+)$/.exec(s)
                 if (m) {
                     if (mul != 1)
                         this.directiveError(lf("multiplication not supported with saved stacks"));
                     if (this.stackpointers.hasOwnProperty(m[1]))
+                        // TODO: ARM-specific???
```
Comment (mmoskal) [inline pxtlib/emitter/assembler.ts]: Not ARM-specific

File: pxtlib/emitter/assembler.ts [hunk 3]
```diff
@@ -323,18 +316,22 @@ namespace ts.pxtc.assembler {
...
                         v = 4 * (this.stack - this.stackpointers[m[1]] + parseInt(m[2]))
                     else
                         this.directiveError(lf("saved stack not found"))
                 }
 
+                // TODO: ARM-specific???
```
Comment (mmoskal) [inline pxtlib/emitter/assembler.ts]: specific to CPUs which can only load 8 bit of constant data from an instruction at a time - I think this includes AVR - the @hi and @lo get the first and second byte of the address of label relative to the beginning of the assembly file

File: pxtlib/emitter/assembler.ts [hunk 4]
```diff
@@ -390,17 +387,14 @@ namespace ts.pxtc.assembler {
                 if (this.finalEmit)
                     this.directiveError(lf("unknown label: {0}", name));
                 else
+                    // Is this a joke?
```
Comment (mmoskal) [inline pxtlib/emitter/assembler.ts]: There needs to be a value here, and in the first emit labels might not be resolved yet. I think there was some reason not to use 0

File: pxtlib/emitter/backbase.ts [hunk 5]
```diff
@@ -0,0 +1,671 @@
...
+    // ARM, so a mapping will be needed for other processors
+
+    // Assumptions:
+    //
+    // - special registers include: pc, lr
+    // - registers are 32-bit (really???)
+    // - pop/push receive register sets: { r0, r3, r4 }
+    // - r0 is the current value (from expression evaluation)
+    // - also used (fixed) are r1, r2, r5, r6 (what about r3, r4?)
+    // - registers for runtime calls (r1,r2,r3)
```
Comment (mmoskal) [inline pxtlib/emitter/backbase.ts]: also r0

File: pxtlib/emitter/backbase.ts [hunk 6]
```diff
@@ -0,0 +1,671 @@
...
+    // that said, the assumptions below about registers are based on
+    // ARM, so a mapping will be needed for other processors
+
+    // Assumptions:
+    //
+    // - special registers include: pc, lr
+    // - registers are 32-bit (really???)
+    // - pop/push receive register sets: { r0, r3, r4 }
+    // - r0 is the current value (from expression evaluation)
+    // - also used (fixed) are r1, r2, r5, r6 (what about r3, r4?)
```
Comment (mmoskal) [inline pxtlib/emitter/backbase.ts]: r5 is for captured locals in lambda and r6 for globals; it's unclear why not r4 and r5

File: pxtlib/emitter/hexfile.ts [hunk 7]
```diff
@@ -0,0 +1,525 @@
...
+        function swapBytes(str: string) {
+            let r = ""
+            let i = 0
+            for (; i < str.length; i += 2)
+                r = str[i] + str[i + 1] + r
+            assert(i == str.length)
+            return r
+        }
+
+        // TODO: generalize and lift out ARM-specific code
```
Comment (mmoskal) [inline pxtlib/emitter/hexfile.ts]: I think there's nothing ARM-specific in this function

File: cli/cli.ts (hunk 1 above)
Comment (tballmsft) [inline cli/cli.ts]: done.

File: pxtlib/emitter/assembler.ts (hunk 2 above)
Comment (tballmsft) [inline pxtlib/emitter/assembler.ts]: thx

File: pxtlib/emitter/assembler.ts (hunk 3 above)
Comment (tballmsft) [inline pxtlib/emitter/assembler.ts]: thx

File: pxtlib/emitter/assembler.ts (hunk 4 above)
Comment (tballmsft) [inline pxtlib/emitter/assembler.ts]: ok

File: test/avr/.vscode/launch.json [hunk 8]
```diff
@@ -0,0 +1,77 @@
+{
```
Comment (pelikhan) [inline test/avr/.vscode/launch.json]: we already have a "tests" folder, can you move all the ``/test/avr/`` files under there?

File: test/avr/convert.py [hunk 9]
```diff
@@ -0,0 +1,183 @@
+import re
```
Comment (pelikhan) [inline test/avr/convert.py]: pxt is fundamentally a JS project. We should avoid adding dependencies to Python. Can this be rewritten in JS instead?

File: test/avr/.vscode/launch.json (hunk 8 above)
Comment (pelikhan) [inline test/avr/.vscode/launch.json]: Wrong button.

**Pull request merged**